File: 2019/day16.go
package main

import (
	"github.com/lukechampine/advent/utils"
)

var input = utils.ReadInput("day16_input.txt")

func phase(vec []int) []int {
	prod := make([]int, len(vec))
	for stride := 1; stride <= len(vec); stride++ {
		var sum int
		for j := stride - 1; j < len(vec); {
			for k := 0; k < stride && j < len(vec); k++ {
				sum += vec[j]
				j++
			}
			j += stride
			for k := 0; k < stride && j < len(vec); k++ {
				sum -= vec[j]
				j++
			}
			j += stride
		}
		prod[stride-1] = utils.Abs(sum) % 10
	}
	return prod
}

func addPhase(off int, vec []int) []int {
	for i := len(vec) - 2; i >= off-1; i-- {
		vec[i] += vec[i+1]
		vec[i+1] = utils.Abs(vec[i+1]) % 10
	}
	return vec
}

func main() {
	// part 1
	ds := utils.Digits(input)
	for p := 0; p < 100; p++ {
		ds = phase(ds)
	}
	for _, d := range ds[:8] {
		utils.Print(d)
	}
	utils.Println()

	// part 2
	ds = make([]int, len(ds)*10000)

	offset := utils.Atoi(input[:7])
	digs := utils.Digits(input)
	for i := range ds {
		ds[i] = digs[i%len(digs)]
	}
	for p := 0; p < 100; p++ {
		ds = addPhase(offset, ds)
	}
	for _, d := range ds[offset:][:8] {
		utils.Print(d)
	}
	utils.Println()
}

File: 2019/day16.zig
const std = @import("std");
const utils = @import("utils.zig");

fn phase(vec: []i64) []i64 {
    var prod = utils.alloc(i64, vec.len);
    var stride: usize = 1;
    while (stride <= vec.len) : (stride += 1) {
        var sum: i64 = 0;
        var j = stride - 1;
        while (j < vec.len) {
            var k: usize = 0;
            while (k < stride and j < vec.len) : (k += 1) {
                sum += vec[j];
                j += 1;
            }
            j += stride;
            k = 0;
            while (k < stride and j < vec.len) : (k += 1) {
                sum -= vec[j];
                j += 1;
            }
            j += stride;
        }
        prod[stride - 1] = @rem(utils.abs(sum), 10);
    }
    return prod;
}

fn abs(x: i64) i64 {
    return std.math.absInt(x) catch unreachable;
}

fn addPhase(off: usize, vec: []i64) void {
    var i: usize = vec.len - 2;
    while (i >= off - 1) : (i -= 1) {
        vec[i] += vec[i + 1];
        vec[i + 1] = @rem(abs(vec[i + 1]), 10);
    }
}

fn printDigits(ds: []i64) void {
    for (ds) |c| {
        utils.print(c);
    }
    utils.print("\n");
}

pub fn main() void {
    const input = utils.readFile("day16_input.txt");
    var ds = utils.alloc(i64, input.len);
    for (ds) |_, i| {
        ds[i] = utils.parseInt(i64, input[i .. i + 1]);
    }
    // part 1
    var p: usize = 0;
    while (p < 100) : (p += 1) {
        ds = phase(ds);
    }
    printDigits(ds[0..8]);

    // part 2
    var ds10k = utils.alloc(i64, ds.len * 10000);
    var offset = utils.parseInt(usize, input[0..7]);
    for (ds10k) |_, i| {
        var di = i % ds.len;
        ds10k[i] = utils.parseInt(i64, input[di .. di + 1]);
    }
    p = 0;
    while (p < 100) : (p += 1) {
        addPhase(offset, ds10k);
    }
    printDigits(ds10k[offset .. offset + 8]);
}
